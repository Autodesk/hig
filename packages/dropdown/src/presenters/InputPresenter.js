import React, { Component } from "react";
import Icon from "@hig/icon";
import { TextFieldPresenter } from "@hig/text-field";
import "@hig/icon/build/index.css";
import "@hig/text-field/build/index.css";
import PropTypes from "prop-types";

import "./InputPresenter.scss";

export default class InputPresenter extends Component {
  static propTypes = {
    /**
     * Prevents user actions on the field
     */
    disabled: PropTypes.bool,
    /**
     * Error text for the field. Setting this value applies error styling to the entire component.
     */
    errors: PropTypes.string,
    /**
     * Allows managing field's focus via prop. Setting this value overrides the internal focus state management.
     */
    focused: PropTypes.bool,
    /**
     * Prevents user actions on the field
     */
    hasTextEntry: PropTypes.bool,
    /**
     * When true, displays passed error text. When false, displays instructions with error styling.
     */
    hideInstructionsOnErrors: PropTypes.bool,
    /**
     * HTML ID attribute
     */
    id: PropTypes.string,
    /**
     * Icon element that precedes the input.
     */
    icon: PropTypes.node,
    /**
     * Instructional text for the field
     */
    instructions: PropTypes.string,
    /**
     * Text describing what the field represents
     */
    label: PropTypes.string,
    /**
     * Name of the field when submitted with a form
     */
    name: PropTypes.string,
    /**
     * Called when user moves focus from the field
     */
    onBlur: PropTypes.func,
    /**
     * Called after user changes the value of the field
     */
    onChange: PropTypes.func,
    /**
     * Called when user clicks the clear button
     */
    onClearButtonClick: PropTypes.func,
    /**
     * Called when user puts focus onto the field
     */
    onFocus: PropTypes.func,
    /**
     * Called as user changes the value of the field
     */
    onInput: PropTypes.func,
    /**
     * Example of what the user should type into the field
     */
    placeholder: PropTypes.string,
    /**
     * Text describing why the field is required
     */
    required: PropTypes.string,
    /**
     * When true, causes the clear button to appear
     */
    showClearButton: PropTypes.bool,
    /**
     * Value of the field
     */
    value: PropTypes.string
  };

  render() {
    let textFieldPresenter;
    const { hasTextEntry, ...inputProps } = this.props;

    if (hasTextEntry) {
      textFieldPresenter = <TextFieldPresenter {...inputProps} type="text" />;
    } else {
      textFieldPresenter = (
        <TextFieldPresenter {...inputProps} type="button" readOnly />
      );
    }

    return (
      <div className="hig__dropdown__input-wrapper">
        {textFieldPresenter}
        <span className="hig__dropdown__input-caret">
          {/* @TODO: there are variations of the TextField with multiple icons at the end of the input. These icon nodes should be passed as props to TextField. */}
          <Icon name="caret" />
        </span>
      </div>
    );
  }
}
