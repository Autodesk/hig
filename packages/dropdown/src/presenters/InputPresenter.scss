@import "../tokens.scss";

.hig__dropdown__input-wrapper {
  position: relative;

  // TODO: Refactor TextField to avoid needing to override styles
  .hig__text-field__input {
    font-weight: 600;
  }

  // TODO: Refactor TextField to avoid needing to override styles
  .hig__text-field__extra--dropdown-caret {
    bottom: 6px;
  }
}

.hig__dropdown__input-caret {
  position: absolute;
  // 30px = the height of the icon element
  top: calc(#{$menu-top-offset} - 30px);
  right: 7px;
  z-index: -1;

  .hig__dropdown--disabled & {
    opacity: 0.3;
  }
}
