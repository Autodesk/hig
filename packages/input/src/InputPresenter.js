import React from "react";
import PropTypes from "prop-types";
import { css } from "emotion";

import { ThemeContext } from "@hig/themes";

import stylesheet from "./stylesheet";

export default function InputPresenter({
  disabled,
  hasFocus,
  hasHover,
  value,
  onFocus,
  onBlur,
  onMouseEnter,
  onMouseLeave,
  onChange
}) {
  return (
    <ThemeContext.Consumer>
      {({ themeData }) => {
        const styles = stylesheet(
          { isDisabled: disabled, hasFocus, hasHover },
          themeData
        );

        return (
          <div className={css(styles.inputWrapper)}>
            <input
              className={css(styles.input)}
              value={value}
              disabled={disabled}
              onFocus={onFocus}
              onBlur={onBlur}
              onMouseEnter={onMouseEnter}
              onMouseLeave={onMouseLeave}
              onChange={onChange}
            />
            <div className={css(styles.halo)} />
          </div>
        );
      }}
    </ThemeContext.Consumer>
  );
}

InputPresenter.propTypes = {
  disabled: PropTypes.bool,
  hasFocus: PropTypes.bool,
  hasHover: PropTypes.bool,
  value: PropTypes.string,
  onFocus: PropTypes.func,
  onBlur: PropTypes.func,
  onMouseEnter: PropTypes.func,
  onMouseLeave: PropTypes.func,
  onChange: PropTypes.func
};
