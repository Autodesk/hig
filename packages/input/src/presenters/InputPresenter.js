import React from "react";
import PropTypes from "prop-types";
import { cx, css } from "emotion";

import { ThemeContext } from "@hig/theme-context";

import stylesheet from "./InputPresenter.stylesheet";
import {
  availableTagNames,
  availableVariants,
  tagNames,
  variants
} from "../constants";

export default function InputPresenter(props) {
  const {
    disabled,
    hasFocus,
    hasHover,
    inputRef,
    onBlur,
    onFocus,
    onMouseEnter,
    onMouseLeave,
    tagName,
    stylesheet: customStylesheet,
    variant,
    ...otherProps
  } = props;
  return (
    <ThemeContext.Consumer>
      {({ resolvedRoles, metadata }) => {
        const styles = stylesheet(
          { isDisabled: disabled, hasFocus, hasHover, variant },
          resolvedRoles
        );
        const cssStyles = customStylesheet
          ? customStylesheet(
              styles,
              props,
              resolvedRoles,
              metadata.colorSchemeId
            )
          : styles;

        let className = css(cssStyles.input);
        const { className: customClassName } = otherProps;
        if (customClassName) {
          className = cx(
            className,
            variant === variants.PLAIN
              ? customClassName
              : customClassName
                  .split(" ")
                  .reduce((acc, cur) => cx(acc, `${cur.trim()}__input`), "")
          );
        }

        const tagNameKey = tagName.toUpperCase();
        const Element = tagNames[tagNameKey];

        return (
          <Element
            {...otherProps}
            className={className}
            disabled={disabled}
            onBlur={onBlur}
            onFocus={onFocus}
            onMouseEnter={onMouseEnter}
            onMouseLeave={onMouseLeave}
            ref={inputRef}
          />
        );
      }}
    </ThemeContext.Consumer>
  );
}

InputPresenter.propTypes = {
  disabled: PropTypes.bool,
  hasFocus: PropTypes.bool,
  hasHover: PropTypes.bool,
  onBlur: PropTypes.func,
  onFocus: PropTypes.func,
  onMouseEnter: PropTypes.func,
  onMouseLeave: PropTypes.func,
  stylesheet: PropTypes.func,
  tagName: PropTypes.oneOf(availableTagNames),
  variant: PropTypes.oneOf(availableVariants)
};

InputPresenter.defaultProps = {
  tagName: tagNames.INPUT
};
