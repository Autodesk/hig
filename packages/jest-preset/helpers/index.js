/* eslint-disable global-require */
module.exports = {
  takeSnapshotsOf: require("./takeSnapshotsOf")
};
