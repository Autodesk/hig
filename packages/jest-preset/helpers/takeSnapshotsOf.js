const React = require("react");
const renderer = require("react-test-renderer");

module.exports = function takeSnapshotsOf(Component, cases, rendererOptions) {
  cases.forEach(({ desc, props }) => {
    it(desc, () => {
      const vDom = React.createElement(Component, props);
      const tree = renderer.create(vDom, rendererOptions).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });
};
