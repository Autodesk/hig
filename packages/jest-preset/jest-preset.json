{
  "collectCoverageFrom": ["**/src/**/*.{js,jsx}"],
  "coverageDirectory": "<rootDir>/coverage",
  "coveragePathIgnorePatterns": [
    "__gemini__",
    "__stories__",
    "docs",
    "examples/redux",
    "node_modules",
    "packages/jest-preset",
    "packages/vanilla",
    "src/playground"
  ],
  "coverageThreshold": {
    "global": {
      "branches": 50,
      "functions": 50,
      "lines": 50,
      "statements": 50
    }
  },
  "moduleNameMapper": {
    "\\.(css|scss|svg)$": "@hig/jest-preset/fileMock.js"
  },
  "modulePaths": ["node_modules"],
  "moduleFileExtensions": ["js", "jsx", "json"],
  "setupFiles": ["raf/polyfill"],
  "setupTestFrameworkScriptFile": "@hig/jest-preset/setupTests.js",
  "testPathIgnorePatterns": [
    "docs",
    "packages/babel-preset",
    "packages/jest-preset",
    "packages/vanilla"
  ],
  "unmockedModulePathPatterns": ["node_modules"]
}
