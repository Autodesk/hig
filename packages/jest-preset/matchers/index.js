require("./toBeConstants");
require("./toImplementHIGInterfaceOf");
require("./toHavePropTypesAndDocgenInfo");
