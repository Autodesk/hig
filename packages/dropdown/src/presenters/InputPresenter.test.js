import { takeSnapshotsOf } from "@hig/jest-preset/helpers";
import InputPresenter from "./InputPresenter";

describe("Dropdown/presenters/InputPresenter", () => {
  const rendererOptions = {
    createNodeMock(element) {
      return document.createElement(element.type);
    }
  };
  const cases = [
    {
      desc: "renders without props",
      props: {}
    },
    {
      desc: "renders with all props (that dropdown uses)",
      props: {
        type: "button",
        readOnly: true,
        id: "id",
        label: "label",
        instructions: "instructions",
        placeholder: "placeholder",
        disabled: true,
        required: "this is required",
        onBlur: () => {},
        onFocus: () => {},
        onClick: () => {}
      }
    }
  ];

  takeSnapshotsOf(InputPresenter, cases, rendererOptions);
});
