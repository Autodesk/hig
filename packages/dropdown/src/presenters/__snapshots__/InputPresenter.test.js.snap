// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown/presenters/InputPresenter renders with all props (that dropdown uses) 1`] = `
.emotion-4 {
  position: relative;
}

.emotion-0 {
  background-color: transparent;
  box-sizing: border-box;
  border: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  padding-top: 4px;
  padding-right: 0;
  padding-bottom: 4px;
  padding-left: 0;
  outline: none;
  font-size: 14px;
  font-family: ArtifaktElement,sans-serif;
  font-weight: 600;
  line-height: 1.4;
  width: 100%;
  color: #3c3c3c;
  height: calc(36px - 2px);
  -webkit-transition-property: color;
  transition-property: color;
  -webkit-transition-duration: 0.3s;
  transition-duration: 0.3s;
  cursor: pointer;
}

.emotion-0::-webkit-input-placeholder {
  color: rgba(60,60,60,0.4);
}

.emotion-0::-moz-placeholder {
  color: rgba(60,60,60,0.4);
}

.emotion-0:-ms-input-placeholder {
  color: rgba(60,60,60,0.4);
}

.emotion-0::placeholder {
  color: rgba(60,60,60,0.4);
}

.emotion-0::-ms-input-placeholder {
  color: rgba(60,60,60,0.4);
}

.emotion-0::selection {
  background-color: rgba(128,128,128,0.2);
}

.emotion-3 {
  fill: #808080;
  pointer-events: none;
}

.emotion-3 > * {
  fill: #808080;
  pointer-events: none;
}

.emotion-2 {
  -webkit-box-flex: 2;
  -webkit-flex-grow: 2;
  -ms-flex-positive: 2;
  flex-grow: 2;
  position: relative;
  border-style: solid;
  border-color: transparent;
  border-width: 1px;
  border-bottom-color: rgba(128,128,128,0.35);
  border-bottom-style: solid;
  border-bottom-width: 1px;
  opacity: 0.4;
}

.emotion-1 {
  position: absolute;
  top: calc(100% + 1px);
  left: -1px;
  right: 0;
  width: calc(100% + 2px);
}

<div
  className="emotion-4"
>
  <div
    className="emotion-2"
  >
    <input
      className="emotion-0"
      disabled={true}
      id="id"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      placeholder="placeholder"
      readOnly={true}
      required="this is required"
      type="button"
    />
    <div
      className="emotion-1"
    />
  </div>
  <svg
    className="emotion-3"
    color="#808080"
    height="10px"
    style={
      Object {
        "opacity": "0.4",
        "pointerEvents": "none",
        "position": "absolute",
        "right": "0",
        "top": "50%",
        "transform": "translateY(-50%)",
        "transition": "transform 0.3s, color 0.3s",
      }
    }
    viewBox="0 0 10 10"
    width="10px"
    xmlns="http://www.w3.org/2000/svg"
  >
    <title>
      caret-down-m
    </title>
    <path
      d="M8.71 3.22a.75.75 0 0 0-1.06 0L5 5.87 2.35 3.22a.75.75 0 0 0-1.06 0 .74.74 0 0 0 0 1.06l3.36 3.36a.5.5 0 0 0 .7 0l3.36-3.36a.74.74 0 0 0 0-1.06z"
    />
  </svg>
</div>
`;

exports[`Dropdown/presenters/InputPresenter renders without props 1`] = `
.emotion-4 {
  position: relative;
}

.emotion-2 {
  -webkit-box-flex: 2;
  -webkit-flex-grow: 2;
  -ms-flex-positive: 2;
  flex-grow: 2;
  position: relative;
  border-style: solid;
  border-color: transparent;
  border-width: 1px;
  border-bottom-color: rgba(128,128,128,0.35);
  border-bottom-style: solid;
  border-bottom-width: 1px;
}

.emotion-1 {
  position: absolute;
  top: calc(100% + 1px);
  left: -1px;
  right: 0;
  width: calc(100% + 2px);
  height: 0;
  background-color: rgba(128,128,128,0.15);
  -webkit-transition-property: height,color;
  transition-property: height,color;
  -webkit-transition-duration: 0.3s,0.3s;
  transition-duration: 0.3s,0.3s;
}

.emotion-0 {
  background-color: transparent;
  box-sizing: border-box;
  border: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  padding-top: 4px;
  padding-right: 0;
  padding-bottom: 4px;
  padding-left: 0;
  outline: none;
  font-size: 14px;
  font-family: ArtifaktElement,sans-serif;
  font-weight: 600;
  line-height: 1.4;
  width: 100%;
  color: #3c3c3c;
  height: calc(36px - 2px);
  -webkit-transition-property: color;
  transition-property: color;
  -webkit-transition-duration: 0.3s;
  transition-duration: 0.3s;
  cursor: pointer;
}

.emotion-0::-webkit-input-placeholder {
  color: rgba(60,60,60,0.4);
}

.emotion-0::-moz-placeholder {
  color: rgba(60,60,60,0.4);
}

.emotion-0:-ms-input-placeholder {
  color: rgba(60,60,60,0.4);
}

.emotion-0::placeholder {
  color: rgba(60,60,60,0.4);
}

.emotion-0::-ms-input-placeholder {
  color: rgba(60,60,60,0.4);
}

.emotion-0::selection {
  background-color: rgba(128,128,128,0.2);
}

.emotion-3 {
  fill: #808080;
  pointer-events: none;
}

.emotion-3 > * {
  fill: #808080;
  pointer-events: none;
}

<div
  className="emotion-4"
>
  <div
    className="emotion-2"
  >
    <input
      className="emotion-0"
      disabled={undefined}
      onBlur={[Function]}
      onFocus={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      readOnly={true}
    />
    <div
      className="emotion-1"
    />
  </div>
  <svg
    className="emotion-3"
    color="#808080"
    height="10px"
    style={
      Object {
        "opacity": 1,
        "pointerEvents": "none",
        "position": "absolute",
        "right": "0",
        "top": "50%",
        "transform": "translateY(-50%)",
        "transition": "transform 0.3s, color 0.3s",
      }
    }
    viewBox="0 0 10 10"
    width="10px"
    xmlns="http://www.w3.org/2000/svg"
  >
    <title>
      caret-down-m
    </title>
    <path
      d="M8.71 3.22a.75.75 0 0 0-1.06 0L5 5.87 2.35 3.22a.75.75 0 0 0-1.06 0 .74.74 0 0 0 0 1.06l3.36 3.36a.5.5 0 0 0 .7 0l3.36-3.36a.74.74 0 0 0 0-1.06z"
    />
  </svg>
</div>
`;
